tile-service/src/main/java/com/oculusinfo/tile/rest/translation/TileTranslationServiceImpl.java

```diff
@@ -0,0 +1,122 @@
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+package com.oculusinfo.tile.rest.translation;
+
+
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.net.URL;
+import java.util.Properties;
+
+import javax.net.ssl.HttpsURLConnection;
+
+import com.google.inject.Singleton;
+import com.oculusinfo.tile.rest.config.ConfigException;
+import com.oculusinfo.tile.rest.config.ConfigPropertiesUtil;
+
+import org.json.JSONException;
+import org.json.JSONObject;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+
+/**
+ * A service that will translate given text to a target language based on the parameters 
+ * 	passed in.  Currently only supports the Google Translate Service. 
+ *
+ */
+@Singleton
+public class TileTranslationServiceImpl implements TileTranslationService {
+	
+	private static final Logger LOGGER = LoggerFactory.getLogger(TileTranslationServiceImpl.class);
+	
+	public static final String TRANSLATE_API_KEY = "translation.api.key";
+
+    /* (non-Javadoc)
+	 * @see TileUtilsServiceImpl#getTranslationGoogle(JSONObject query)
+	 */
+	public JSONObject getTranslation( JSONObject query ) {
+    	// get the translation arguments from the query
+		JSONObject result = null;
+    	try {
+    		// as we integrate more translation service we can add a more sophisticated selection mechanism
+    		String service = query.getString("service");
+    		if ( service.equals( "google" ) ) {
+    			// get google translate params
+    			String text = query.getString("text");
+        		String target = query.getString("target");
+        		
+        		ConfigPropertiesUtil configUtil = new ConfigPropertiesUtil();
+            	Properties properties = configUtil.getConfigProperties();
+        		String translationApiKey = properties.getProperty(TRANSLATE_API_KEY);
+        		
+            	result = translateGoogle( text, target, translationApiKey );
+            }	
+    	} catch ( JSONException e ) {
+    		LOGGER.error( "Incorrect Configuration for Translation API", e );
+		} catch ( ConfigException e ) {
+			LOGGER.error( "Error with internal configuration properties", e );
+		} 
+		return result;
+	}
+       
+	/*
+	 * Translates the given text using the Google Translate API
+	 */
+    private JSONObject translateGoogle( String text, String target, String key ) {
+		JSONObject result = null;
+		try {
+			// assumes that the text has already been uri encoded
+            String urlStr = "https://www.googleapis.com/language/translate/v2?key=" + key + "&q=" + text + "&target=" + target;
```

Comment:
I'll add the translation.api.endpoint to the default configuration
properties file.

Text and target are provided by the client.  Normally the REST call will
package the strings up and unpack them when we get the params using the
decoder.  I encountered problems with the decoder not handling some tweets
with all the special character encoding so in the client I explicitly
encode the text as a url component before sending.

On Wed, Jul 22, 2015 at 10:42 AM, Daniela Baron <notifications@github.com>
wrote:

> In
> tile-service/src/main/java/com/oculusinfo/tile/rest/translation/TileTranslationServiceImpl.java
> <https://github.com/unchartedsoftware/aperture-tiles/pull/287#discussion_r35220091>
> :
>
> > +    	} catch ( JSONException e ) {
> > +    		LOGGER.error( "Incorrect Configuration for Translation API", e );
> > +		} catch ( ConfigException e ) {
> > +			LOGGER.error( "Error with internal configuration properties", e );
> > +		}
> > +		return result;
> > +	}
> > +
> > +	/*
> > +	 * Translates the given text using the Google Translate API
> > +	 */
> > +    private JSONObject translateGoogle( String text, String target, String key ) {
> > +		JSONObject result = null;
> > +		try {
> > +			// assumes that the text has already been uri encoded
> > +            String urlStr = "https://www.googleapis.com/language/translate/v2?key=" + key + "&q=" + text + "&target=" + target;
>
> Might want to make the url to google also configurable in case they change
> it. In fact might want to make it more generic, something like
> translationServiceEndpoint rather than specifically google, in case an app
> wants to use some other service provider for translation.
>
> Are "text" and "target" provided by the client as part of api request?
> Wondering if any sanitizing required before constructing the url.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/unchartedsoftware/aperture-tiles/pull/287/files#r35220091>
> .
>